windows/hid.c

```diff
@@ -1031,25 +1105,20 @@ DWORD WINAPI hid_internal_notify_callback(HCMNOTIFICATION notify, PVOID context,
 
 	if (device) {
 		/* Call the notifications for the device */
-		struct hid_hotplug_callback *hotplug_cb = hid_hotplug_context.hotplug_cbs;
-		while (hotplug_cb != NULL) {
-			if ((hotplug_cb->events & hotplug_event) &&
-				hid_internal_match_device_id(device->vendor_id, device->product_id, hotplug_cb->vendor_id, hotplug_cb->product_id)) {
-				struct hid_hotplug_callback* cur_hotplug_cb = hotplug_cb;
-				hotplug_cb = cur_hotplug_cb->next;
-
-				if ((*cur_hotplug_cb->callback)(cur_hotplug_cb->handle, device, hotplug_event, cur_hotplug_cb->user_data)) {
-					hid_hotplug_deregister_callback(cur_hotplug_cb->handle);
-
-					/* Last callback was unregistered */
-					if (hid_hotplug_context.hotplug_cbs == NULL) {
-						break;
-					}
+		struct hid_hotplug_callback **current = &hid_hotplug_context.hotplug_cbs;
+		while (*current) {
+			struct hid_hotplug_callback *callback = *current;
+			if ((callback->events & hotplug_event) && hid_internal_match_device_id(device->vendor_id, device->product_id, callback->vendor_id, callback->product_id)) {
+				int result = (callback->callback)(callback->handle, device, hotplug_event, callback->user_data);
+				/* If the result is non-zero, we remove the callback and proceed */
+				/* Do not use the deregister call as it locks the mutex, and we are currently in a lock */
+				if (result) {
+					*current = (*current)->next;
+					free(callback);
+					continue;
 				}
 			}
-			else {
-				hotplug_cb = hotplug_cb->next;
-			}
+			current = &callback->next;
```

Comment:
Here you are using the `callback` which potentially freed already I would rather save its next value beforehand,
also shouldn't the `hid_internal_hotplug_cleanup` called here?

Reply:
Actually, no, if `free` is called then the execution never reaches this point, as `continue` will jump it to the beginning of the cycle.